src/item.cpp

```diff
@@ -197,25 +197,15 @@ item::item(JsonObject &jo)
     deserialize(jo);
 }
 
-item::~item()
-{
-}
-
-void item::make( const std::string new_type, bool scrub )
+item& item::convert( const itype_id& new_type )
 {
     type = find_type( new_type );
-    contents.clear();
-
-    if (scrub) {
-        components.clear();
-        charges = -1;
-        bday = 0;
-        name = "";
-        curammo = NULL;
+    if( is_null() ) {
```

Comment:
> As far as I know, C++ doesn't have strict exception handling enforcement rules like Java

I don't know much about Java but in C++ they just propagate upwards until they reach a `catch` statement otherwise `std::terminate` is called as a last resort. I don't think you can force a caller to handle an exception.

> Having a null-object serves as a suitable alternative.

Indeed it works but it's arguably a poorer alternative as it partly breaks encapsulation. In the below example that caller doesn't need to know anything about the workings of `do_everything()` to report an error.

```
void do_everything() {
    if (... ) {
        throw std::exception( "failed because of foo" );
    } else if (...) {`
        throw std::exception( "failed because of quux" );
    }
}
try {
    do_something();
} catch ( ...) {
    debugmsg ...
}
```

But if we rely on checking a return value we have to break the task into separate steps and possibly back out of them:

```
bool do_part_foo() {
   if (... ) {
       return false;
   }
}
bool do_part_quux() {
   if (... ) {
       return false;
   }
}

if (doo_part_foo() == false ) {
    debugmsg("can't do quux");
}
if (doo_part_foo() == false ) {
    cleanup_from_foo();
    debugmsg("can't do quux");
}
```

> the changes would be quite massive 

We could start by wrapping `main()` in a `try/catch` statement and throw exceptions only for truly exceptional circumstances that should never occur, a missing item definition would be a good example. No other changes are required to existing code and new code can if it wishes choose to handle exceptions if there's a rational solution.

The only net loss would be that the older code will now terminate with a `debugmsg` for a missing item definition rather than continuing (which might be most appropriate in many cases anyway). The net gain is that the new code is much cleaner.